File: src/main/java/mediathek/gui/tabs/tab_film/DownloadSubtitleAction.java
package mediathek.gui.tabs.tab_film;

import mediathek.config.Konstanten;
import mediathek.mainwindow.MediathekGui;
import mediathek.tool.FileDialogs;
import mediathek.tool.FileUtils;
import mediathek.tool.PathExtensions;
import mediathek.tool.SwingErrorDialog;
import mediathek.tool.subtitles.MVSubtitle;
import mediathek.tool.subtitles.detector.TimedTextFormatDetector;
import mediathek.tool.subtitles.ttml2.AssExporter;
import mediathek.tool.subtitles.ttml2.SubRipHtmlExporter;
import mediathek.tool.subtitles.ttml2.Ttml2Parser;
import mediathek.tool.subtitles.vtt.WebVttToTtml2Converter;

import javax.swing.*;
import java.awt.event.ActionEvent;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

public class DownloadSubtitleAction extends AbstractAction {
    private final GuiFilme guiFilme;

    public DownloadSubtitleAction(GuiFilme guiFilme) {
        this.guiFilme = guiFilme;
        putValue(Action.NAME, "Untertitel-Datei sofort laden...");
    }

    @Override
    public void actionPerformed(ActionEvent e) {
        guiFilme.getCurrentlySelectedFilm().ifPresent(film -> {
            var selectedFile = FileDialogs.chooseSaveFileLocation(MediathekGui.ui(), "Untertitel speichern", "");
            if (selectedFile != null) {
                Path tempSubtitleFile = null;
                var selectedFilePath = selectedFile.toPath();

                try {
                    tempSubtitleFile = FileUtils.downloadToTempFile(film.getSubtitleUrl());
                    MVSubtitle.moveWithFallback(tempSubtitleFile, selectedFilePath);

                    var res = TimedTextFormatDetector.detect(selectedFilePath, true);
                    if (!res.valid()) {
                        //invalid result
                        JOptionPane.showMessageDialog(MediathekGui.ui(), "Untertitelformat konnte nicht erkannt werden.",
                                Konstanten.PROGRAMMNAME, JOptionPane.ERROR_MESSAGE);
                    }
                    else {
                        //valid result
                        if (res.format() == TimedTextFormatDetector.Format.UNKNOWN) {
                            JOptionPane.showMessageDialog(MediathekGui.ui(), "Untertitelformat wird nicht unterstützt.",
                                    Konstanten.PROGRAMMNAME, JOptionPane.ERROR_MESSAGE);
                            return;
                        }

                        selectedFilePath = MVSubtitle.addFileExtension(selectedFilePath, res.format());

                        if (res.format() == TimedTextFormatDetector.Format.WEBVTT) {
                            //convert WebVTT to TTML2
                            WebVttToTtml2Converter converter = new WebVttToTtml2Converter();
                            var newPath = PathExtensions.withExtension(selectedFilePath, ".ttml");
                            converter.convert(selectedFilePath, newPath);
                            selectedFilePath = newPath;
                        }

                        Ttml2Parser parser = new Ttml2Parser();
                        var ttmlDoc = parser.parse(selectedFilePath);

                        //SRT conversion
                        var srtPath = PathExtensions.withExtension(selectedFilePath, ".srt");
                        var srtStr = new SubRipHtmlExporter().export(ttmlDoc);
                        Files.writeString(srtPath, srtStr);

                        //ASS conversion
                        var assPath = PathExtensions.withExtension(selectedFilePath, ".ass");
                        var assOptions = new AssExporter.Options(384, 288, false);
                        var assStr = new AssExporter(assOptions).export(ttmlDoc);
                        Files.writeString(assPath, assStr);

                        JOptionPane.showMessageDialog(MediathekGui.ui(), "Vorgang erfolgreich.",
                                Konstanten.PROGRAMMNAME, JOptionPane.WARNING_MESSAGE);
                    }
                }
                catch (Exception ex) {
                    SwingErrorDialog.showExceptionMessage(MediathekGui.ui(),
                            "Untertitel konnte nicht geladen werden.", ex);
                }
                finally {
                    try {
                        if (tempSubtitleFile != null)
                            Files.deleteIfExists(tempSubtitleFile);
                    }
                    catch (IOException ignored) {
                    }
                }
            }
            else
                JOptionPane.showMessageDialog(MediathekGui.ui(), "Vorgang wurde abgebrochen.",
                        Konstanten.PROGRAMMNAME, JOptionPane.WARNING_MESSAGE);
        });
    }
}

File: src/main/kotlin/mediathek/gui/tabs/tab_film/actions/DownloadSubtitleAction.kt
/*
 * Copyright (c) 2026 derreisende77.
 * This code was developed as part of the MediathekView project https://github.com/mediathekview/MediathekView
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package mediathek.gui.tabs.tab_film.actions

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.swing.Swing
import mediathek.config.Konstanten
import mediathek.daten.DatenFilm
import mediathek.tool.FileDialogs
import mediathek.tool.SwingErrorDialog
import mediathek.tool.subtitles.SubtitleExportResult
import mediathek.tool.subtitles.SubtitleExportService
import java.awt.Frame
import java.awt.event.ActionEvent
import java.util.*
import java.util.function.Supplier
import javax.swing.AbstractAction
import javax.swing.Action
import javax.swing.JOptionPane

class DownloadSubtitleAction(
    private val owner: Frame,
    private val currentlySelectedFilm: Supplier<Optional<DatenFilm>>,
) : AbstractAction() {
    private val uiScope = CoroutineScope(SupervisorJob() + Dispatchers.Swing)

    init {
        putValue(Action.NAME, "Untertitel-Datei sofort laden...")
    }

    override fun actionPerformed(e: ActionEvent?) {
        val film = currentlySelectedFilm.get().orElse(null) ?: return
        val selectedFile = FileDialogs.chooseSaveFileLocation(owner, "Untertitel speichern", "")

        if (selectedFile == null) {
            JOptionPane.showMessageDialog(
                owner,
                "Vorgang wurde abgebrochen.",
                Konstanten.PROGRAMMNAME,
                JOptionPane.WARNING_MESSAGE
            )
            return
        }

        isEnabled = false
        uiScope.launch {
            try {
                val result = SubtitleExportService.downloadAndExport(film.subtitleUrl, selectedFile.toPath())

                when (result) {
                    SubtitleExportResult.InvalidFormat -> {
                        JOptionPane.showMessageDialog(
                            owner,
                            "Untertitelformat konnte nicht erkannt werden.",
                            Konstanten.PROGRAMMNAME,
                            JOptionPane.ERROR_MESSAGE
                        )
                    }

                    SubtitleExportResult.UnsupportedFormat -> {
                        JOptionPane.showMessageDialog(
                            owner,
                            "Untertitelformat wird nicht unterstützt.",
                            Konstanten.PROGRAMMNAME,
                            JOptionPane.ERROR_MESSAGE
                        )
                    }

                    is SubtitleExportResult.Success -> {
                        JOptionPane.showMessageDialog(
                            owner,
                            buildCompletionMessage(result),
                            Konstanten.PROGRAMMNAME,
                            if (result.failures.isEmpty()) JOptionPane.INFORMATION_MESSAGE else JOptionPane.WARNING_MESSAGE
                        )
                    }

                    is SubtitleExportResult.Failure -> {
                        SwingErrorDialog.showExceptionMessage(
                            owner,
                            "Untertitel konnte nicht geladen werden.",
                            result.exception
                        )
                    }
                }
            } finally {
                isEnabled = true
            }
        }
    }

    private fun buildCompletionMessage(result: SubtitleExportResult.Success): String {
        val successLine = "Erfolgreich erstellt: ${result.successes.joinToString(", ")}."
        if (result.failures.isEmpty()) {
            return successLine
        }

        val failureLine = result.failures.keys.joinToString(", ")
        return "$successLine\nFehlgeschlagen: $failureLine."
    }

}
